docs/items/consumables.md

```diff
@@ -0,0 +1,369 @@
+
+A list of vanilla effects can be found in `ConsumeEffect`.
+
+Every `ConsumeEffect` has two methods: `getType`, which specifies the registry object `ConsumeEffect.Type`; and `apply`, which is called on the item has been fully consumed. `apply` takes in three arguments: the current `Level` the player is in, the `ItemStack` the consumable was called on, and the `LivingEntity` consuming the object. When the effect is successfully applied, the method returns `true`, or `false` if it failed.
+
+A `ConsumeEffect` can be created by implementing the interface and [registering] the `ConsumeEffect.Type` with the associated `MapCodec` and `StreamCodec` to `BuiltInRegistries#CONSUME_EFFECT_TYPE`:
+
+```java
+public record UsePortalConsumeEffect(ResourceKey<Level> level)
+    implements ConsumeEffect, Portal {
+
+    @Override
+    public boolean apply(Level level, ItemStack stack, LivingEntity entity) {
+        if (entity.canUsePortal(false)) {
+            entity.setAsInsidePortal(this, entity.blockPosition());
+
+            // Can successfully use portal
+            return true;
+        }
+
+        // Cannot use portal
+        return false;
+    }
+
+    @Override
+    public ConsumeEffect.Type<? extends ConsumeEffect> getType() {
+        // Set to registered object
+        return USE_PORTAL.get();
+    }
+
+    @Override
+    @Nullable
+    public TeleportTransition getPortalDestination(ServerLevel level, Entity entity, BlockPos pos) {
+        // Set teleport location
+    }
+}
+
+// In some registrar class
+// Assume there is some DeferredRegister<ConsumeEffect.Type<?>> CONSUME_EFFECT_TYPES
+public static final Supplier<ConsumeEffect.Type<UsePortalConsumeEffect>> USE_PORTAL =
+    CONSUME_EFFECT_TYPES.register("use_portal", () -> new ConsumeEffect.Type<>(
+        ResourceKey.codec(Registries.DIMENSION).optionalFieldOf("dimension")
+            .xmap(UsePortalConsumeEffect::new, UsePortalConsumeEffect::level),
+        ResourceKey.streamCodec(Registries.DIMENSION)
+            .map(UsePortalConsumeEffect::new, UsePortalConsumeEffect::level)
+    ));
+
+// For some Item.Properties that is adding a CONSUMABLE component
+Consumable.builder()
+    .onConsume(
+        new UsePortalConsumeEffect(Level.END)
+    )
+    .build();
+```
+
+### `ItemUseAnimation`
+
+`ItemUseAnimation` is functionally an enum which doesn't define anything besides its id and name. It is only when used in conjunction with `ItemHandRenderer#renderArmWithItem` for first person and `PlayerRenderer#getArmPose` for third person does any actual rendering changes take place. As such, simply creating a new `ItemUseAnimation` will only function similarly to `ItemUseAnimation#NONE`.
+
+To apply some animation, you need to implement `IClientItemExtensions#applyForgeHandTransform` for first person and/or `IClientItemExtensions#getArmPose` for third person rendering.
```

Comment:
I think it makes sense to extract `IClientExtensions`, and generally all client extensions into their own article(s), but that can be left for a future PR. For right now, I think the mention here is good enough. I don't want to link to BEWLR since although they use the same interface, how its used is different.

Reply:
Fair enough.